File: frontend/GameHubFront/src/Games/batalhaRural/helpers/TokenInfo.ts
import { Utils } from "./Utils";
import { Token } from "../entities/Token";
import { CoordinateHelper } from "./CoordinateHelper";
import { Direction } from "../enums/Direction";
import { Axis } from "../enums/Axis";

export class TokenInfo {
    readonly baseOnePosX: number;
    readonly baseOnePosY: number;
    readonly tokenSize: number;
    readonly tokenDir: Direction;
    readonly tokenName: string;
    readonly tokenChar: string;
    
    get baseOneFinalPosX(): number {
        return CoordinateHelper.calcHorizontalFinalPosition(this.baseOnePosX, this.tokenSize, this.tokenDir);
    }

    get baseOneFinalPosY(): number {
        return CoordinateHelper.calcVerticalFinalPosition(this.baseOnePosY, this.tokenSize, this.tokenDir);
    }

    get baseZeroFinalPosX(): number {
        return Utils.toBaseZero(this.baseOneFinalPosX);
    }

    get baseZeroFinalPosY(): number {
        return Utils.toBaseZero(this.baseOneFinalPosY);
    }

    get baseZeroPosX(): number {
        return Utils.toBaseZero(this.baseOnePosX);
    }

    get baseZeroPosY(): number {
        return Utils.toBaseZero(this.baseOnePosY);
    }

    get tokenAxis(): Axis {
        if (this.tokenDir == Direction.Up || this.tokenDir == Direction.Down) { return Axis.Y; }
        else { return Axis.X; }
    }

    constructor(token: Token) {
        this.tokenSize = token.size;
        this.tokenDir = token.direction;
        this.tokenName = token.name;
        this.tokenChar = token.character;
        this.baseOnePosX = token.positionX;
        this.baseOnePosY = token.positionY;
    }
}

File: backend/gameHubBack/services/BatalhaRural/TokenInfo.cs
using gameHubBack.Entities.BatalhaRural;
using gameHubBack.Enums.BatalhaRural;

namespace gameHubBack.Services.BatalhaRural;

public class TokenInfo
{
    public int BaseOnePosX { get; }
    public int BaseOnePosY { get; }
    public int TokenSize { get; }
    public Direction TokenDir { get; }
    public string TokenName { get; }
    public string TokenChar { get; }

    public int BaseOneFinalPosX => CoordinateHelper.CalcHorizontalFinalPosition(BaseOnePosX, TokenSize, TokenDir);
    public int BaseOneFinalPosY => CoordinateHelper.CalcVerticalFinalPosition(BaseOnePosY, TokenSize, TokenDir);
    public int BaseZeroFinalPosX => BaseOneFinalPosX - 1;
    public int BaseZeroFinalPosY => BaseOneFinalPosY - 1;
    public int BaseZeroPosX => BaseOnePosX - 1;
    public int BaseZeroPosY => BaseOnePosY - 1;

    public Axis TokenAxis => TokenDir is Direction.Up or Direction.Down ? Axis.Y : Axis.X;

    public TokenInfo(GameToken token)
    {
        TokenSize = TokenCatalog.GetSize(token.Type);
        TokenDir = token.Direction;
        TokenName = TokenCatalog.GetName(token.Type);
        TokenChar = TokenCatalog.GetCharacter(token.Type);
        BaseOnePosX = token.PositionX;
        BaseOnePosY = token.PositionY;
    }
}